src/DependencyInjection/Compiler/MenuBuilderPass.php

```diff
@@ -44,6 +45,38 @@ public function process(ContainerBuilder $container)
                 $menuBuilders[$attributes['alias']] = [$id, $attributes['method']];
             }
         }
+
+
+        foreach ($container->findTaggedServiceIds('knp_menu.menu_builder_provider') as $id => $tags) {
+            $def = $container->getDefinition($id);
+
+            if (!$def->isPublic()) {
+                throw new \InvalidArgumentException(sprintf('Menu builder services must be public but "%s" is a private service.', $id));
+            }
+
+            if ($def->isAbstract()) {
+                throw new \InvalidArgumentException(sprintf('Abstract services cannot be registered as menu builders but "%s" is.', $id));
+            }
+
+            // We must assume that the class value has been correctly filled, even if the service is created by a factory
```

Comment:
why must we assume that the `class` is configured if this is a factory created service?

Reply:
because we need to call the static method on the class.

Note that Symfony's EventSubscriber does the same (this comment is a copy/paste of the Symfony one actually)

Reply:
ah right, the `$class::` below... 
ok, now i get it.